frontend/src/modules/auth/actions.ts

```diff
@@ -0,0 +1,116 @@
+import jwtDecode from "jwt-decode";
+
+import api from "../../utils/api";
+import { createNetworkActionTypes, networkAction } from "../../utils/actions";
+import { selectAuthState } from "../../selectors"
+import { isNullish } from "../../utils/functions"
+
+
+export const LOG_OUT = "AUTH.LOG_OUT";
+
+export const PERFORM_AUTH = createNetworkActionTypes("AUTH.PERFORM");
+export const REFRESH_AUTH_TOKEN = createNetworkActionTypes("AUTH.REFRESH_TOKEN");
+
+export const logOut = () => ({ type: LOG_OUT });
+
+const decodeUserID = tokens => ({ tokens, currentUserID: jwtDecode(tokens.access).user_id })
+
+export const performAuth = (username, password) => async (dispatch, getState) => {
+
+    if (getState().auth.isFetching) return false;
+
+    await dispatch(
+        networkAction(
+            PERFORM_AUTH,
+            api.auth.token({ username, password }),
+            { transform: decodeUserID }));
+
+}
+
+/**
+ * This function checks to see if a user is logged in and has access and refresh tokens.
+ * If so, it checks how long it will be before the access token expires. If it is due to
+ * expire then it will try to refresh the token, as long as the refresh token hasn't
+ * also expired.
+ * 
+ * The function will dispatch the LOG_OUT action if it is unable to refresh the tokens.
+ * 
+ * It returns true if the access token is currently valid and can be used to make
+ * api requests, and false otherwise.
+ * 
+ * @returns boolean
+ */
+export const refreshAuthToken = () => async (dispatch, getState) => {
+    // Refresh the access token if it is within 2 minutes of expiring
+    const TIME_BEFORE_TOKEN_EXPIRES = 2 * 60    
```

Comment:
This should be somewhere centralized so we can find it. Also in order to have this not be synchronized to the summary calls we might want to make both timers be dependants and add an offset of a few seconds to make so they won't trigger at the same time.

Reply:
Also this may need to be bigger with templates taking 2-3 minutes to submit it's likely the app will miss the window if a template submission is timed badly.

Reply:
Submitting a template doesn't block the UI thread, so the timer isn't affected by that 2-3 minute wait, and we can refresh the token while we are waiting for the submission response. It wouldn't cause a problem in the frontend, but I'm not sure what would happen in the backend. If the backend received the submission request and was spending several minutes processing it, and in the meantime refreshed the access token would it "cancel" the submit operation because the token associated with the submit request had expired? I have no idea, but I'll change this to 5 minutes just in case.

Reply:
There's no good place to put that `TIME_BEFORE_TOKEN_EXPIRES` constant right now. I would worry about that later when have time to implement a better system for managing auth.